app/lib/report_helper.rb

```diff
@@ -6,30 +6,39 @@ def external_report_info(report, view_level, params)
     data[:highest_tax_counts] = htc
     data[:taxonomy_details] = td
     data[:view_level] = view_level
+    data[:all_categories] = all_categories
     data
   end
 
   def report_details(report)
+    if report
     {
       report_info: report,
       pipeline_info: report.pipeline_output,
       sample_info: report.pipeline_output.sample,
       project_info: report.pipeline_output.sample.project,
       background_model: report.background
     }
+    else
+      {}
+    end
   end
 
   def compute_taxon_zscores(report)
-    summary = report.background.summarize
-    total_reads = report.pipeline_output.total_reads
-    pipeline_output_id = report.pipeline_output.id
-    data = TaxonCount.connection.select_all("select tax_id, pipeline_output_id, tax_level, count, name, count_type from taxon_counts where pipeline_output_id = #{pipeline_output_id}").to_hash
-    data.each do |h|
-      h[:rpm] = compute_rpm(h["count"], total_reads)
+    if report
+      summary = report.background.summarize
+      total_reads = report.pipeline_output.total_reads
+      pipeline_output_id = report.pipeline_output.id
+      data = TaxonCount.connection.select_all("select tax_id, pipeline_output_id, tax_level, count, name, count_type from taxon_counts where pipeline_output_id = #{pipeline_output_id}").to_hash
+      data.each do |h|
+        h[:rpm] = compute_rpm(h["count"], total_reads)
+      end
+      data_and_background = (data + summary).group_by { |h| [h["tax_id"], h["tax_level"], h["name"], h["count_type"]] }.map { |_k, v| v.reduce(:merge) }.select { |h| h["count"] }
+      zscore_array = data_and_background.map { |h| { tax_id: h["tax_id"], tax_level: h["tax_level"], name: h["name"], rpm: h[:rpm], hit_type: h["count_type"], zscore: compute_zscore(h[:rpm], h[:mean], h[:stdev]) } }
+      zscore_array
+    else
+      []
```

Comment:
When you have a long function like that it's acceptable to put a `return [] unless report` at the top.  In this case that would be better than indenting the entire long function.

Note it is still considered poor form to return from the middle of a long function.